scheds/rust/scx_rustlite/src/main.rs

```diff
@@ -0,0 +1,314 @@
+        let nr_kernel_dispatches = self.skel.bss().nr_kernel_dispatches as u64;
+
+        info!(
+            "nr_enqueues={} nr_user_dispatched={} nr_kernel_dispatches={}",
+            nr_enqueues, nr_user_dispatches, nr_kernel_dispatches
+        );
+        log::logger().flush();
+    }
+
+    fn run(&mut self, shutdown: Arc<AtomicBool>) -> Result<()> {
+        let mut prev_ts = SystemTime::now();
+
+        while !shutdown.load(Ordering::Relaxed) && self.read_bpf_exit_kind() == 0 {
+            let curr_ts = SystemTime::now();
+            let elapsed = curr_ts
+                .duration_since(prev_ts)
+                .unwrap_or_else(|_| Duration::from_secs(0));
+
+            self.schedule();
+
+            // Print scheduler statistics every second
+            if elapsed > Duration::from_secs(1) {
+                self.print_stats();
+                prev_ts = curr_ts;
+            }
+        }
+        self.report_bpf_exit_kind()
+    }
+}
+
+impl<'a> Drop for Scheduler<'a> {
+    fn drop(&mut self) {
+        if let Some(struct_ops) = self.struct_ops.take() {
+            drop(struct_ops);
+        }
+        info!("Unregister {} scheduler", SCHEDULER_NAME);
+    }
+}
+
+// Set scheduling class for the scheduler itself to SCHED_EXT
+fn use_sched_ext() -> i32 {
+    let pid = std::process::id();
+    let param: sched_param = sched_param { sched_priority: 0 };
+    let res = unsafe { sched_setscheduler(pid as i32, SCHED_EXT, &param as *const sched_param) };
+    res
+}
+
+fn main() -> Result<()> {
+    let opts = Opts::parse();
+
+    let loglevel = simplelog::LevelFilter::Info;
+
+    let mut lcfg = simplelog::ConfigBuilder::new();
+    lcfg.set_time_level(simplelog::LevelFilter::Error)
+        .set_location_level(simplelog::LevelFilter::Off)
+        .set_target_level(simplelog::LevelFilter::Off)
+        .set_thread_level(simplelog::LevelFilter::Off);
+    simplelog::TermLogger::init(
+        loglevel,
+        lcfg.build(),
+        simplelog::TerminalMode::Stderr,
+        simplelog::ColorChoice::Auto,
+    )?;
+
+    // Make sure to use the SCHED_EXT class at least for the scheduler itself.
+    let res = use_sched_ext();
+    if res != 0 {
+        bail!("Failed to all sched_setscheduler: {}", res);
+    }
+
+    let mut sched = Scheduler::init(&opts)?;
+    let shutdown = Arc::new(AtomicBool::new(false));
+    let shutdown_clone = shutdown.clone();
+    ctrlc::set_handler(move || {
+        shutdown_clone.store(true, Ordering::Relaxed);
+    })
+    .context("Error setting Ctrl-C handler")?;
+
+    /*
+     * Spawn a heartbeat thread to periodically trigger the check to run the user-space scheduler.
+     *
+     * Without this thread we may starve the scheduler task if the system is completely idle and
+     * hit the sched-ext watchdog that would auto-kill this scheduler.
+     */
+    thread::spawn(|| loop {
+        thread::sleep(Duration::from_secs(1));
+    });
```

Comment:
We could also do this by setting a timer in BPF code, but this works too

Reply:
Yeah, a bpf_timer is definitely nicer, I'll add that.